src/scala/org/pantsbuild/zinc/compiler/OutputUtils.scala

```diff
@@ -0,0 +1,116 @@
+/**
+ * Copyright (C) 2012 Typesafe, Inc. <http://www.typesafe.com>
+ */
+
+package org.pantsbuild.zinc.compiler
+
+import java.io.File
+import java.nio.file.attribute.BasicFileAttributes
+import java.nio.file.{FileVisitResult, Files, Path, Paths, SimpleFileVisitor}
+import java.util.jar.{JarEntry, JarInputStream, JarOutputStream}
+import scala.annotation.tailrec
+import scala.collection.mutable
+import scala.util.Try
+
+object OutputUtils {
+
+  /**
+   * Sort the contents of the `dir` in lexicographic order.
+   *
+   * @param dir File handle containing the contents to sort
+   * @return sorted set of all paths within the `dir`
+   */
+  def sort(dir:File): mutable.TreeSet[Path] = {
+    val sorted = new mutable.TreeSet[Path]()
+
+    val fileSortVisitor = new SimpleFileVisitor[Path]() {
+      override def preVisitDirectory(path: Path, attrs: BasicFileAttributes): FileVisitResult = {
+        sorted.add(path)
+        FileVisitResult.CONTINUE
+      }
+
+      override def visitFile(path: Path, attrs: BasicFileAttributes): FileVisitResult = {
+        sorted.add(path)
+        FileVisitResult.CONTINUE
+      }
+    }
+
+    Files.walkFileTree(dir.toPath, fileSortVisitor)
+    sorted
+  }
+
+  def relativize(base: String, path: Path): String = {
+    new File(base).toURI().relativize(new File(path.toString).toURI()).getPath()
+  }
+
+  /**
+   * Create a JAR of of filePaths provided.
+   *
+   * @param filePaths set of all paths to be added to the JAR
+   * @param outputJarPath Absolute Path to the output JAR being created
+   * @param jarEntryTime time to be set for each JAR entry
+   */
+  def createJar(
+    base: String, filePaths: mutable.TreeSet[Path], outputJarPath: Path, jarEntryTime: Long) {
+
+    val target = new JarOutputStream(Files.newOutputStream(outputJarPath))
+
+    def addToJar(source: Path, entryName: String): FileVisitResult = {
+      val jarEntry = new JarEntry(entryName)
+      // setting jarEntry time to a fixed value for all entries within the jar so that jars are
+      // byte-for-byte reproducible.
+      jarEntry.setTime(jarEntryTime)
+
+      target.putNextEntry(jarEntry)
+      Files.copy(source, target)
```

Comment:
Hm... it looks like you're properly collecting directories above (ie, collecting directories in pre-visit), but I feel like this method will fail if the `source: Path` is a directory? It's important to actually store directory entries in the zip file, so I think that this case needs to add the jarEntry for a directory as a directory, and then skip appending the content.

You might consider having the FileVisitor store a tuple of `(path: Path, isFile: Boolean)` (which can still be sorted) to avoid needing to check whether the path is a directory here.

Between this and my comment in the test, I think the test needs an update to create a classfile under a directory.